File: backup/skip-gram.py
import sys

def group(seq, n):
	return (seq[i:i+n] for i in range(len(seq)-n+1))

data = None
if(len(sys.argv) >= 2):
	data = sys.argv[1].readlines()
else:
	data = sys.stdin.readlines()

SKIP = 10

def customprint(x):
	[a,b] = [' '.join(y.split('\t')[2:]).rstrip() for y in [x[0], x[2]]]
	print a + '\t' + b + '\t' + x[1].split(' ')[-1].rstrip()

g = filter(lambda x: True if "NOP" in x[1] and int(x[1].split(' ')[-1]) < SKIP else False, list(group(data,3)))
map(customprint , g)

File: process/skip-gram.rb
# Preparse data for the n-gram model (each "gram" being a S-V-O action)
# Allows skipgrams up to SKIP

SKIP = 10
LEN = 2 # n for n-gram

magic = 2*(LEN-1)+1

ARGF.each_line.each_cons(magic) do |group|
  if !(group.first =~ /NOP/)
    parsed = group.map { |x| x =~ /NOP/ ? x.split(" ").last.to_i : x }
    if parsed.select { |x| x.is_a? Integer }.map { |x| x < SKIP }.reduce(:&)
      strings = parsed.select { |x| x.is_a? String }.map do |x|
        x.split("\t").join(" ").strip
    	end
      puts strings.join("\t") if strings.uniq.size > 1
    end
  end
end
